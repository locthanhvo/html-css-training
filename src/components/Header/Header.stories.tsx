import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Header from '.';

export default {
  title: 'Components/Header',
  component: Header,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof Header>;

export const HeaderComponent: Story = {
  args: { width: '1140px' },
};
