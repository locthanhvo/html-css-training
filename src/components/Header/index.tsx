import { memo, useMemo } from 'react';
import {
  Link,
  useLocation,
  useMatches,
  useParams,
  useRouteLoaderData,
} from 'react-router-dom';

import {
  Box,
  Breadcrumb,
  BreadcrumbItem,
  Flex,
  Heading,
  Text,
} from '@chakra-ui/react';

// Utils
import { formatBreadcrumb } from '@/utils';

// Icons
import { BreadcrumbIcon, NotifyIcon } from '@/components/Icons';

const Header = ({ ...props }) => {
  const { pathname } = useLocation();
  const matches = useMatches();
  const param = useParams();

  const { title } =
    (useRouteLoaderData(matches.at(-1)?.id ?? pathname) as {
      title?: string;
    }) || {};

  const renderBreadcrumb = useMemo(() => {
    const basePath = Object.values(param).reduce<string>(
      (path, param) => path.replace('/' + param, ''),
      pathname,
    );
    const paths = formatBreadcrumb(basePath);

    return paths.map(({ pathName, breadcrumbName }) => {
      const colorText = `/${pathName}` === pathname ? 'blue.300' : 'gray.500';
      return (
        <BreadcrumbItem key={pathName}>
          <Link to={pathName}>
            <Text fontSize="xs" color={colorText}>
              {breadcrumbName.replace(/\b\w/g, (l) => l.toUpperCase())}
            </Text>
          </Link>
        </BreadcrumbItem>
      );
    });
  }, [param, pathname]);

  return (
    <>
      <Flex
        bgColor="primary"
        maxW="full"
        px={10}
        py={7}
        justifyContent="space-between"
        {...props}
      >
        <Box>
          <Breadcrumb spacing="8px" separator={<BreadcrumbIcon />}>
            {renderBreadcrumb}
          </Breadcrumb>
        </Box>

        <Flex gap={3} alignItems="center">
          <NotifyIcon />
        </Flex>
      </Flex>

      <Heading py={5} px={10} as="h1" fontSize="lg" fontWeight="semibold">
        {title}
      </Heading>
    </>
  );
};

export default memo(Header);
