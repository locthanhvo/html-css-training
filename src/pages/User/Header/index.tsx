import Button from '@components/Button'
import Filter from '@components/Filter'
import Heading from '@components/Heading'
import AddIcon from '@components/Icons/AddIcon'
import SearchInput from '@components/SearchInput'
import Text from '@components/Text'
import { Colors, FontSize, FontWeight } from '@themes'

interface HeaderProps {
  onChangeSearch: (e: React.ChangeEvent<HTMLInputElement>) => void
  onClickSelectOption: (option: string | React.ChangeEvent<HTMLSelectElement>) => void
  selected: string
  onOpenForm: (e: React.FormEvent<Element>) => void
  onDeleteMutiple: () => void
  checked: number
}

const Header = ({
  onChangeSearch,
  onClickSelectOption,
  onOpenForm,
  onDeleteMutiple,
  selected,
  checked,
}: HeaderProps) => {
  return (
    <div className='d-flex main-header'>
      <Heading content='Team members' tag='h1' />
      <div className='d-flex main-tools'>
        <div className='d-flex  main-tool-group'>
          <div className='display-button'>
            <Button variants='danger' onClick={onDeleteMutiple} isDisabled={checked === 0}>
              <Text
                content='Delete All'
                color={Colors.White}
                fontSize={FontSize.TwoSmall}
                fontWeight={FontWeight.Bold}
              />
            </Button>
          </div>
          <SearchInput onChangeSearch={onChangeSearch} />
        </div>
        <div className='d-flex main-tool-group'>
          <Filter
            onClickSelectOption={onClickSelectOption}
            selectedOption={selected}
            checked={checked}
            onDeleteMutiple={onDeleteMutiple}
          />
          <Button variants='primary' onClick={onOpenForm}>
            <Text
              content='Add User'
              color={Colors.White}
              fontSize={FontSize.TwoSmall}
              fontWeight={FontWeight.Bold}
            />
            <AddIcon />
          </Button>
        </div>
      </div>
    </div>
  )
}

export default Header
