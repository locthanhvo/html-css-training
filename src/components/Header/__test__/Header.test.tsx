import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Header from '..';

const mockUseMatches = { at: jest.fn() };
const mockUseRouteLoaderData = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useMatches: () => mockUseMatches,
  useRouteLoaderData: () => mockUseRouteLoaderData,
}));

describe('Header Component', () => {
  it('Match Header component', () => {
    const element = render(
      <BrowserRouter>
        <Header width="1140px" />
      </BrowserRouter>,
    );

    expect(element).toMatchSnapshot();
  });
});
