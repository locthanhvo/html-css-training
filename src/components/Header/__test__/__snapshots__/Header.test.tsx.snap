// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Match Header component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-ghqtji"
      >
        <div
          class="css-0"
        >
          <nav
            aria-label="breadcrumb"
            class="chakra-breadcrumb css-0"
          >
            <ol
              class="chakra-breadcrumb__list css-70qvj9"
            />
          </nav>
        </div>
        <div
          class="css-1plht7o"
        >
          <svg
            fill="none"
            height="22"
            viewBox="0 0 20 22"
            width="20"
          >
            <path
              d="M16 7C16 5.4087 15.3679 3.88258 14.2426 2.75736C13.1174 1.63214 11.5913 1 10 1C8.4087 1 6.88258 1.63214 5.75736 2.75736C4.63214 3.88258 4 5.4087 4 7C4 14 1 16 1 16H19C19 16 16 14 16 7Z"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M11.73 20C11.5542 20.3031 11.3019 20.5547 10.9982 20.7295C10.6946 20.9044 10.3504 20.9965 10 20.9965C9.64964 20.9965 9.30541 20.9044 9.00179 20.7295C8.69818 20.5547 8.44583 20.3031 8.27002 20"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
      <h1
        class="chakra-heading css-1y338v4"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="css-ghqtji"
    >
      <div
        class="css-0"
      >
        <nav
          aria-label="breadcrumb"
          class="chakra-breadcrumb css-0"
        >
          <ol
            class="chakra-breadcrumb__list css-70qvj9"
          />
        </nav>
      </div>
      <div
        class="css-1plht7o"
      >
        <svg
          fill="none"
          height="22"
          viewBox="0 0 20 22"
          width="20"
        >
          <path
            d="M16 7C16 5.4087 15.3679 3.88258 14.2426 2.75736C13.1174 1.63214 11.5913 1 10 1C8.4087 1 6.88258 1.63214 5.75736 2.75736C4.63214 3.88258 4 5.4087 4 7C4 14 1 16 1 16H19C19 16 16 14 16 7Z"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M11.73 20C11.5542 20.3031 11.3019 20.5547 10.9982 20.7295C10.6946 20.9044 10.3504 20.9965 10 20.9965C9.64964 20.9965 9.30541 20.9044 9.00179 20.7295C8.69818 20.5547 8.44583 20.3031 8.27002 20"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <h1
      class="chakra-heading css-1y338v4"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
